service/iot-service-samples/azure-sas-credential-sample/src/main/java/samples/com/microsoft/azure/sdk/iot/AzureSasCredentialSample.java

```diff
@@ -0,0 +1,359 @@
+import com.microsoft.azure.sdk.iot.service.exceptions.IotHubException;
+import com.microsoft.azure.sdk.iot.service.jobs.JobClient;
+import com.microsoft.azure.sdk.iot.service.jobs.JobClientOptions;
+import com.microsoft.azure.sdk.iot.service.jobs.JobResult;
+import lombok.extern.slf4j.Slf4j;
+
+import javax.crypto.Mac;
+import javax.crypto.spec.SecretKeySpec;
+import java.io.IOException;
+import java.io.UnsupportedEncodingException;
+import java.net.URLEncoder;
+import java.nio.charset.Charset;
+import java.nio.charset.StandardCharsets;
+import java.security.InvalidKeyException;
+import java.security.NoSuchAlgorithmException;
+import java.time.Instant;
+import java.util.Base64;
+import java.util.UUID;
+
+/**
+ * Sample that demonstrates how to use the {@link AzureSasCredential} authentication mechanism when using the service clients.
+ * This sample also provides a helper function that demonstrates how to generate shared access signatures that can be
+ * used to create instances of {@link AzureSasCredential}.
+ */
+@Slf4j
+public class AzureSasCredentialSample
+{
+    public static void main(String[] args)
+    {
+        SamplesArguments parsedArguments = new SamplesArguments(args);
+
+        String sharedAccessSignature = parsedArguments.getSharedAccessSignature();
+
+        // Shared access signatures can be generated by the SDK for you by using the service client constructors
+        // that take in your IoT Hub's connection string. This sample will demonstrate how to construct service clients
+        // without giving them the connection string by giving them shared access signatures instead.
+
+        // This sample includes a helper function that shows how to build shared access signatures below.
+        // Search for "generateSharedAccessSignature"
+        //
+        // More details on how to build this shared access signature can be found here:
+        // https://docs.microsoft.com/en-us/azure/iot-hub/iot-hub-devguide-security#security-token-structure
+        AzureSasCredential credential = new AzureSasCredential(sharedAccessSignature);
+
+        // "my-azure-iot-hub.azure-devices.net" for example
+        String iotHubHostName = parsedArguments.getIotHubHostName();
+
+        String newDeviceId = runRegistryManagerSample(iotHubHostName, credential);
+
+        runTwinClientSample(iotHubHostName, credential, newDeviceId);
+
+        runServiceClientSample(iotHubHostName, credential, newDeviceId);
+
+        // The AzureSasCredential object supports updating the shared access signature which is important to do
+        // once the previous signature has expired, or is close to expiring. This sample assumes that no renewal will
+        // be necessary since the sample completes in a few seconds. After updating the credential object, any
+        // clients that were constructed with that credential will start using the updated signature
+
+        //String updatedSignature = "some new shared access signature";
```

Comment:
did you mean to leave these commented snippets?

Reply:
Yes. This sample doesn't require the user to provide a second shared access signature, but I did want to leave this code in to draw the reader's attention to this feature of AzureSasCredential

Reply:
If you want to draw attention to it, I suggest adding some comments. It is not very clear otherwise. However, I thing what might be more beneficial is showing how they can update it. In this case the renewal is left to the customer.

Reply:
They are! I somehow did not see them.